### PR TITLE
[SOL] Remove `neg` and change the semantics of `sub`

### PR DESCRIPTION
**Pull request by LucasSte**
Following the alterations made in https://github.com/solana-labs/rbpf/pull/489, this PR removes the `neg` instructions and changes the semantics of `sub` when one of the operands is an immediate.

`sub r1, 2` now means `r1 = 2 - r1`, instead of `r1 = r1 - 2`.
`neg r1` is represented as `r1 = 0  - r1`.

This is the second task in https://github.com/solana-labs/solana/issues/34250.

### CODE REVIEW

File: llvm/test/CodeGen/SBF/sub_reversed_immidiate.ll [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+;     vec[0] = 50 - vec[0];
+;     vec[1] = vec[3] - vec[2];
+; }
+;
+; Compilation flag:
+; clang -S -emit-llvm test.c
+
+
+; Function Attrs: noinline nounwind optnone ssp uwtable(sync)
+define void @test_func_64(ptr noundef %vec) #0 {
```
Comment (nvjle) [inline llvm/test/CodeGen/SBF/sub_reversed_immidiate.ll]: All of the new test cases are too verbose/complicated and should simplified. We only need to unit test a single instruction, so the tests should be very narrowly written that way. We want minimal testcases-- a single LLVM IR instruction is all that is needed to cover the current functionality. For example, a test for `sub64` can be succintly written as follows:
```
; Function Attrs: norecurse nounwind readnone
define dso_local i64 @test_sub64_ri(i64 %a) local_unnamed_addr #0 {
entry:
  %t = sub nsw i64 50, %a
; CHECK: sub64 r{{[0-9]+}}, 50
  ret i64 %t
}
```
Note also that I use FileCheck's regex matching on the register name to make it slightly less impervious to register allocation changes over time.

Similarly, write the other test cases as shown above.

Also, you can just fold the negate tests into `sub_reversed_immidiate.ll` and remove `neg_inst_with_sub.ll`.
Finally, please rename `sub_reversed_immidiate.ll` -> `sub_reversed_immediate.ll`.

File: llvm/lib/Target/SBF/SBFInstrInfo.td [hunk 2]
```diff
@@ -306,24 +331,6 @@ let Constraints = "$dst = $src2" in {
     }
 }
```
Comment (nvjle) [inline llvm/lib/Target/SBF/SBFInstrInfo.td:313]: The entire part of the patch above involving the `ALU` multiclass and the `SUB` definition need to be reworked and simplified. That is, there is no need to refactor and duplicate the multiclass. We can succinctly adjust the original multiclass by conditionally swapping the pattern operands as follows:
```
@@ -264,6 +264,12 @@ class ALU_RR<SBFOpClass Class, SBFArithOp Opc,
 }
 
 multiclass ALU<SBFArithOp Opc, string Mnemonic, SDNode OpNode> {
+  // Match swapped operand order only for special case of 'sub reg,imm`.
+  defvar DoSwap = !eq(OpNode, sub);
+  defvar RegImmPat = !if(DoSwap, (OpNode i64immSExt32:$imm, GPR:$src2),
+                                 (OpNode GPR:$src2, i64immSExt32:$imm));
+  defvar RegImmPat32 = !if(DoSwap,(OpNode i32immSExt32:$imm, GPR32:$src2),
+                                  (OpNode GPR32:$src2, i32immSExt32:$imm));
   def _rr : ALU_RR<SBF_ALU64, Opc,
                    (outs GPR:$dst),
                    (ins GPR:$src2, GPR:$src),
@@ -273,7 +279,7 @@ multiclass ALU<SBFArithOp Opc, string Mnemonic, SDNode OpNod
e> {
                    (outs GPR:$dst),
                    (ins GPR:$src2, i64imm:$imm),
                    Mnemonic # "64 $dst, $imm",
-                   [(set GPR:$dst, (OpNode GPR:$src2, i64immSExt32:$imm))]>;
+                   [(set GPR:$dst, RegImmPat)]>;
   def _rr_32 : ALU_RR<SBF_ALU, Opc,
                    (outs GPR32:$dst),
                    (ins GPR32:$src2, GPR32:$src),
@@ -283,7 +289,7 @@ multiclass ALU<SBFArithOp Opc, string Mnemonic, SDNode OpNod
e> {
                    (outs GPR32:$dst),
                    (ins GPR32:$src2, i32imm:$imm),
                    Mnemonic # "32 $dst, $imm",
-                   [(set GPR32:$dst, (OpNode GPR32:$src2, i32immSExt32:$imm))]>
;
+                   [(set GPR32:$dst, RegImmPat32)]>;
 }
 
 let Constraints = "$dst = $src2" in {
@@ -306,24 +312,6 @@ let Constraints = "$dst = $src2" in {
     }
 }
```

Comment (nvjle) [reply llvm/lib/Target/SBF/SBFInstrInfo.td:313]: One related general question I have is-- should we conditionally be generating the swapped special case `sub`? We need to exactly match the runtime semantics on chain today and down the road.

Comment (LucasSte) [reply llvm/lib/Target/SBF/SBFInstrInfo.td:313]: The new syntax can be behind the `sbfv2` cpu perhaps?

Comment (LucasSte) [reply llvm/lib/Target/SBF/SBFInstrInfo.td:313]: I couldn't find a way to encapsulate the new `sub` inside a predicate for `sbfv2`. All the other example cases do not redefine the same instruction. Instead, the predicate encapsulate definitions that only exist for a certain feature and do not override other definitions.

Comment (LucasSte) [reply llvm/lib/Target/SBF/SBFInstrInfo.td:313]: Everything should be behind a feature flag now.

File: llvm/lib/Target/SBF/SBFInstrInfo.td [hunk 3]
```diff
@@ -748,9 +775,13 @@ let Constraints = "$dst = $val" in {
...
 def : Pat<(atomic_load_sub_32 ADDRri:$addr, GPR32:$val),
-          (XFADDW32 ADDRri:$addr, (NEG_32 GPR32:$val))>;
+          (XFADDW32 ADDRri:$addr, (NEG_32 GPR32:$val))>, Requires<[SBFv1]>;
+def : Pat<(atomic_load_sub_64 ADDRri:$addr, GPR:$val),
+          (XFADDD ADDRri:$addr, (NEG_64 GPR:$val))>, Requires<[SBFv1]>;
+def : Pat<(atomic_load_sub_32 ADDRri:$addr, GPR32:$val),
+          (XFADDW32 ADDRri:$addr, (SUB_ri_32 GPR32:$val, 0))>, Requires<[SBFv2]>;
 def : Pat<(atomic_load_sub_64 ADDRri:$addr, GPR:$val),
-          (XFADDD ADDRri:$addr, (NEG_64 GPR:$val))>;
+          (XFADDD ADDRri:$addr, (SUB_ri GPR:$val, 0))>, Requires<[SBFv2]>;
```
Comment (LucasSte) [inline llvm/lib/Target/SBF/SBFInstrInfo.td]: I must use `SUB_ri_rev` and `SUB_ri_32_rev` in these two cases.

File: llvm/lib/Target/SBF/SBFInstrInfo.td (hunk 2 above)
Comment (nvjle) [reply llvm/lib/Target/SBF/SBFInstrInfo.td:313]: Thanks for incorporating my offline suggestions, this is ready to land (other than some trivial coding standard nits, which I'll note shortly).

File: llvm/lib/Target/SBF/SBFInstrInfo.td [hunk 4]
```diff
@@ -263,7 +265,7 @@ class ALU_RR<SBFOpClass Class, SBFArithOp Opc,
   let SBFClass = Class;
 }
 
-multiclass ALU<SBFArithOp Opc, string Mnemonic, SDNode OpNode> {
+multiclass ALU<SBFArithOp Opc, string Mnemonic, SDNode OpNode, bit UseImmPat = 1> {
```
Comment (nvjle) [inline llvm/lib/Target/SBF/SBFInstrInfo.td:268]: LLVM Coding standard nit: (80-column rules) Please move the new parameter to the next line, aligned with the `<`'.

File: llvm/lib/Target/SBF/SBFInstrInfo.td [hunk 5]
```diff
@@ -306,6 +312,19 @@ let Constraints = "$dst = $src2" in {
     }
 }
 
+// Special case for SBFv2
+// In SBFv1, `sub reg, imm` is interpreted as reg = reg - imm,
+// but in SBFv2 it means reg = imm - reg
```
Comment (nvjle) [inline llvm/lib/Target/SBF/SBFInstrInfo.td:317]: LLVM Coding standard nit: (Proper prose, capitalization, punctuation) Please add period/full-stops (as in the original comments that I had written-- but your wording is fine).
```
+// Special case cpu wart:
+// Operand order for 'sub reg,imm' is interpreted
+// differently for pre-sbfv2 cpu versus sbfv2. In the former case, the
+// VM executes 'reg-imm'. For the latter, 'imm-reg'.
+// Define two sets of 'sub' patterns accordingly.
```

File: llvm/lib/Target/SBF/SBFInstrInfo.td [hunk 6]
```diff
@@ -318,12 +337,21 @@ class NEG_RR<SBFOpClass Class, SBFArithOp Opc,
...
-                     [(set GPR:$dst, (ineg i64:$src))]>;
+                     []>;
   def NEG_32: NEG_RR<SBF_ALU, SBF_NEG, (outs GPR32:$dst), (ins GPR32:$src),
                      "neg32 $dst",
-                     [(set GPR32:$dst, (ineg i32:$src))]>;
+                     []>;
 }
 
+// Instruction `neg` exists on SBFv1, but not on SBFv2
+// In SBFv2, the negate operation is done with a subtraction
```
Comment (nvjle) [inline llvm/lib/Target/SBF/SBFInstrInfo.td:347]: Ditto: Periods/full-stops.

**Review by nvjle — APPROVED** [review]

**Pull request merged**